toolchain/check/facet_type.cpp

```diff
@@ -433,21 +405,51 @@ class SubstImplWitnessAccessCallbacks : public SubstInstCallbacks {
       }
     }
 
-    auto SetFullyRewritten(Context& context, SemIR::ImplWitnessAccess access,
-                           SemIR::InstId rewritten_to_inst_id) -> void {
+    auto SetFullyRewritten(Context& context, SemIR::LocId loc_id,
+                           SemIR::ImplWitnessAccess access,
+                           SemIR::InstId access_inst_id,
+                           SemIR::InstId rewritten_to_inst_id)
+        -> SemIR::InstId {
       auto it = map_.find({GetKey(context, access)});
       CARBON_CHECK(it != map_.end());
-      // TODO: If state == FullyRewrtten and the inst id is different (according
-      // to `CompareFacetTypeConstraintValues`), we can diagnose writing two
-      // different values for the same associated constant immediately?
-      //
-      // TODO: Once the above is done, we don't need to do the SortAndDedupe
-      // step in ResolveFacetTypeRewriteConstraints()? We can just convert this
-      // `map_` into a new set of `RewriteConstraint`s which will already be
-      // deduped.
-      if (it->second.state == BeingRewritten) {
+      if (it->second.state == FullyRewritten) {
+        auto ord = CompareFacetTypeConstraintValues(context, it->second.inst_id,
+                                                    rewritten_to_inst_id);
+        if (ord != std::weak_ordering::equivalent) {
+          if (it->second.inst_id != SemIR::ErrorInst::InstId &&
+              rewritten_to_inst_id != SemIR::ErrorInst::InstId) {
+            CARBON_DIAGNOSTIC(AssociatedConstantWithDifferentValues, Error,
+                              "associated constant {0} given two different "
+                              "values {1} and {2}",
+                              InstIdAsConstant, InstIdAsConstant,
+                              InstIdAsConstant);
+            // Use inst id ordering as a simple proxy for source ordering, to
+            // try to name the values in the same order they appear in the facet
+            // type.
+            auto source_order1 =
+                it->second.inst_id.index < rewritten_to_inst_id.index
+                    ? it->second.inst_id
+                    : rewritten_to_inst_id;
+            auto source_order2 =
+                it->second.inst_id.index >= rewritten_to_inst_id.index
+                    ? it->second.inst_id
+                    : rewritten_to_inst_id;
+            // TODO: It would be nice to note the places where the values are
+            // assigned but rewrite constraint instructions are from canonical
+            // constant values, and have no locations. We'd need to store a
+            // location along with them in the rewrite constraints.
+            context.emitter().Emit(
+                loc_id, AssociatedConstantWithDifferentValues, access_inst_id,
+                source_order1, source_order2);
+          }
+          it->second.inst_id = SemIR::ErrorInst::InstId;
+        }
```

Comment:
I think this is too much work to be doing in this setter; also, this can only happen in the call to this from the rewrite loop, not the call within the subst callbacks. Maybe this diagnostic logic could be kept in the rewrite loop?